src/osdc/Objecter.h

```diff
@@ -1250,6 +1250,55 @@ class Objecter : public md_config_obs_t, public Dispatcher {
   // to be drained by consume_blacklist_events.
   bool blacklist_events_enabled;
   std::set<entity_addr_t> blacklist_events;
+  struct pg_mapping_t {
+    std::vector<int> up;
+    int up_primary = -1;
+    std::vector<int> acting;
+    int acting_primary = -1;
+
+    pg_mapping_t() {}
+    pg_mapping_t(std::vector<int> up, int up_primary,
+                 std::vector<int> acting, int acting_primary)
+               : up(up), up_primary(up_primary),
+                 acting(acting), acting_primary(acting_primary) {}
+  };
+  std::shared_mutex pg_mapping_lock;
+  std::map<int64_t, std::map<pg_t, std::pair<epoch_t, pg_mapping_t>>> pg_mappings;
+
+  // convenient accessors
+  bool lookup_pg_mapping(const pg_t& pg, epoch_t e, pg_mapping_t* pg_mapping) {
+    std::shared_lock l{pg_mapping_lock};
+    auto i = pg_mappings.find(pg.pool());
+    if (i == pg_mappings.end())
+      return false;
+    auto j = i->second.find(pg);
+    if (j == i->second.end())
+      return false;
+    if (j->second.first != e) // stale
+      return false;
+    *pg_mapping = j->second.second;
+    return true;
+  }
+  void update_pg_mapping(const pg_t& pg, epoch_t e,
+                         pg_mapping_t&& pg_mapping) {
+    std::lock_guard l{pg_mapping_lock};
+    auto& pool_pg_mapping = pg_mappings[pg.pool()];
+    pool_pg_mapping.insert({pg, {e, std::move(pg_mapping)}});
+  }
+  void prune_pg_mapping(const mempool::osdmap::map<int64_t,pg_pool_t>& pools) {
+    std::lock_guard l{pg_mapping_lock};
+    for (auto it = pg_mappings.begin(); it != pg_mappings.end();) {
+      if (pools.count(it->first)) {
+        // may need to remove pool's non-existent pg_mappings too,
+        // since in nautilus we allow pg_num of pool to shrink.
+        // however we can leave it as it is for now since in general
+        // there is no harm in keeping some stale pg mappings in the cache..
```

Comment:
Yeah.  Or if we took a vector<pg_mapping_t> approach we could just ``resize(pg_num)``.